### ISSUE TITLE
Erroneous report for unused local variable 

### ISSUE THREAD

**Issue by JonRobertson** [integrated-application-development/sonar-delphi#211]
### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.3.0

### Issue description

DelphiLint says "Remove this unused local variable" for the variable pt in the code block below.

`procedure TfrmMain.btnTestClick(Sender: TObject);
begin
  var pt := Point(btnTest.Left, tlbrMain.Top + btnTest.Height);
  pt := ClientToScreen(pt);
  pumnuTest.Popup(pt.X, pt.Y);
end;
`

### Steps to reproduce

Create Delphi VCL project.
Add a button named btnTest and a TPopupMenu named pumnuTest.
Add an OnClick event for btnTest and paste the code above for the event.
Analyze the unit using DelphiLint.

**Comment by fourls** [comment]
Hi @JonRobertson,

I can't seem to reproduce this problem using the steps above. For me, there are two issues raised for the variable, both of which are working correctly:

- Give this short identifier a more meaningful name. (ShortIdentifier)
- Rename this variable to match the expected naming convention. (VariableName)

My unit looks like this, is this correct?

```delphi
unit formMain;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.StdCtrls, Vcl.Menus;

type
  TfrmMain = class(TForm)
    btnTest: TButton;
    pumnuTest: TPopupMenu;
    procedure btnTestClick(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  frmMain: TfrmMain;

implementation

uses System.Types;

{$R *.dfm}

procedure TfrmMain.btnTestClick(Sender: TObject);
begin
  var pt := Point(btnTest.Left, btnTest.Top + btnTest.Height);
  pt := ClientToScreen(pt);
  pumnuTest.Popup(pt.X, pt.Y);
end;

end.
```

**Comment by Cirras** [comment]
Hi @JonRobertson,

Without more information, I will be closing this issue soon (as the behavior cannot be reproduced).
Please provide additional reproduction steps if possible.

**Comment by luebbe** [comment]
Hello @Cirras,

I'd like to reopen this issue. It still exists with the current 1.8 version of sonar-delphi

The following short method triggers the message. NodeData is necessary as a dummy var, because it is passed as var parameter into the function ContainsFile(). In this case it is not used afterwards, which seems to make sonar-delphi think that it is unnecessary.

```pascal
procedure TForm.vstFilesDblClick(Sender: TObject);
var
  NodeData: PNodeData;
begin
  if ContainsFile(vstFiles.FocusedNode, NodeData) then
    DoSomething;
end;
```

![grafik](https://github.com/user-attachments/assets/bf498fff-ead4-4362-bc1e-630f2935efc0)

**Comment by Cirras** [comment]
Hi @luebbe,

I can tell you that the rule **as written** should not be raising an issue on this type of code.

In your case there may be a name/invocation resolution failure on the overall call to `ContainsFile`, causing the reference to local variable `NodeData` to be unresolved. This would trick the rule into thinking that the variable is never referenced.

There's a decent chance that these are different problems, so let's keep them separate.
Please open a new issue with a minimal reproducible case that I can test the analyzer against.

**Comment by JonRobertson** [comment]
@Cirras,

I've tested a few times and cannot get this to reproduce consistently. The "Remove this unused local variable" message appears during one analysis and doesn't during the next. I have only had it appear twice in a couple dozen analysis, with no changes that should affect the related variable.

**Comment by JonRobertson** [comment]
@luebbe and @Cirras 

luebbe, the code you posted can't be easily tested. I suspect you are using TVirualTreeView. As Ciarras suggested, try to create a simple, minimal test scenario. I hacked your code to something that is will fail to run, but sufficient to test the Delphi Sonar analysis. As with my own code, the "Remove this unused local variable" appeared once, disappeared after the next analysis, and I could not get it to appear again with several change/analysis iterations. The code that I tested is pasted below.

Note that neither the Delphi compiler nor Sonar reported a critical issue, which Pascal Analyzer does report:

WARN9: nodeData : PNodeData in uMain.pas(24): Var parameters that are used, but never set

`unit uMain;

interface

uses
  Vcl.Forms;

type
  TForm1 = class(TForm)
    procedure btnTestClick(Sender: TObject);
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

type
  TNodeData = NativeInt;
  PNodeData = ^TNodeData;

function ContainsFile(var nodeData: PNodeData): Boolean;
begin
  NodeData^ := 10;
  Result := True;
end;

procedure TForm1.btnTestClick(Sender: TObject);
var
  NodeData: PNodeData;
begin
  if ContainsFile(NodeData) then
    Exit;
end;

end.`

**Comment by luebbe** [comment]
Using the attached project the error occurs with every call to delphi-lint.

But it really looks like it is related to VirtualTreeView, because the second method which basically does the same thing doesn't show the error. You have to add your path to VirtualTreeView to the project options.

![grafik](https://github.com/user-attachments/assets/124f4ee5-4c5e-4acb-821d-39c23fe668d0)

[UnusedVariableTest.zip](https://github.com/user-attachments/files/16714238/UnusedVariableTest.zip)

**Comment by Cirras** [comment]
> @luebbe and @Cirras
> 
> luebbe, the code you posted can't be easily tested. I suspect you are using TVirualTreeView. As Ciarras suggested, try to create a simple, minimal test scenario. I hacked your code to something that is will fail to run, but sufficient to test the Delphi Sonar analysis. As with my own code, the "Remove this unused local variable" appeared once, disappeared after the next analysis, and I could not get it to appear again with several change/analysis iterations. The code that I tested is pasted below.
> 
> Note that neither the Delphi compiler nor Sonar reported a critical issue, which Pascal Analyzer does report:
> 
> WARN9: nodeData : PNodeData in uMain.pas(24): Var parameters that are used, but never set WARN7: NodeData : PNodeData in uMain.pas(32): Local variables that are referenced before they are set
> 
> ```pascal
> unit uMain;
> 
> interface
> 
> uses
>   Vcl.Forms;
> 
> type
>   TForm1 = class(TForm)
>     procedure btnTestClick(Sender: TObject);
>   end;
> 
> var
>   Form1: TForm1;
> 
> implementation
> 
> {$R *.dfm}
> 
> type
>   TNodeData = NativeInt;
>   PNodeData = ^TNodeData;
> 
> function ContainsFile(var nodeData: PNodeData): Boolean;
> begin
>   nodeData^ := 10;
>   Result := True;
> end;
> 
> procedure TForm1.btnTestClick(Sender: TObject);
> var
>   NodeData: PNodeData;
> begin
>   if ContainsFile(NodeData) then
>     Exit;
> end;
> 
> end.
> ```

Hi @JonRobertson, 

> As with my own code, the "Remove this unused local variable" appeared once, disappeared after the next analysis, and I could not get it to appear again with several change/analysis iterations.

This is an interesting wrinkle - analysis is usually quite deterministic/stable and doesn't have issues flicking on and off like that.
I'm interested in looking into your example further, so I will reopen this issue to track that.

**Comment by Cirras** [comment]
> Note that neither the Delphi compiler nor Sonar reported a critical issue, which Pascal Analyzer does report:
> 
> WARN9: nodeData : PNodeData in uMain.pas(24): Var parameters that are used, but never set WARN7: NodeData : PNodeData in uMain.pas(32): Local variables that are referenced before they are set

Also @JonRobertson, it's worth noting that both of these Pascal Analyzer issues are making an assumption about whether a particular `var` parameter is expected to be initialized 

It's quite possible that the variable is initialized within the called routine, which would be perfectly safe.
However, I would say that `out` is best practice in cases like that.

See #132 for a relevant rule that we're interested in implementing in the future.
PRs are welcome.

**Comment by Cirras** [comment]
Hi @luebbe,

> Using the attached project the error occurs with every call to delphi-lint. RAD Studio 12.1.
>
> But it really looks like it is related to VirtualTreeView, because the second method which basically does the same thing doesn't show the error. You have to add your path to VirtualTreeView to the project options.

Thank you for supplying a reproducible case.
Can you please raise a new issue so that it can be tracked/looked at separately?

**Comment by Cirras** [comment]
Hi @JonRobertson,

Unfortunately, I'm still not able to reproduce with your latest example (using DelphiLint 1.1.1 and SonarDelphi 1.8.0).
I'm going to close the issue again.

On balance, I think it's more likely that there's some kind of DelphiLint bug causing issues to disappear for you. In general, SonarDelphi's analysis results are quite deterministic/stable.

It's been a while since this issue was originally raised.
If you aren't already, I'd recommend getting on to the latest versions of both DelphiLint and SonarDelphi and seeing if the issue is resolved.
